src/core/lib/event_engine/thread_pool.cc

```diff
@@ -20,138 +20,135 @@
 
 #include "src/core/lib/event_engine/thread_pool.h"
 
+#include <memory>
 #include <utility>
 
+#include <grpc/support/log.h>
+
 #include "src/core/lib/gprpp/thd.h"
 
 namespace grpc_event_engine {
 namespace experimental {
 
-ThreadPool::Thread::Thread(ThreadPool* pool)
-    : pool_(pool),
-      thd_(
-          "posix_eventengine_pool",
-          [](void* th) { static_cast<ThreadPool::Thread*>(th)->ThreadFunc(); },
-          this, nullptr, grpc_core::Thread::Options().set_tracked(false)) {
-  thd_.Start();
+void ThreadPool::StartThread(StatePtr state) {
+  state->thread_count.Add();
+  grpc_core::Thread(
+      "event_engine",
+      [](void* arg) {
+        ThreadFunc(*std::unique_ptr<StatePtr>(static_cast<StatePtr*>(arg)));
+      },
+      new StatePtr(state), nullptr,
+      grpc_core::Thread::Options().set_tracked(false).set_joinable(false))
+      .Start();
 }
-ThreadPool::Thread::~Thread() { thd_.Join(); }
-
-void ThreadPool::Thread::ThreadFunc() {
-  pool_->ThreadFunc();
-  // Now that we have killed ourselves, we should reduce the thread count
-  grpc_core::MutexLock lock(&pool_->mu_);
-  pool_->nthreads_--;
-  // Move ourselves to dead list
-  pool_->dead_threads_.push_back(this);
-
-  if (pool_->nthreads_ == 0) {
-    if (pool_->forking_) pool_->fork_cv_.Signal();
-    if (pool_->shutdown_) pool_->shutdown_cv_.Signal();
+
+void ThreadPool::ThreadFunc(StatePtr state) {
+  while (state->queue.Step()) {
   }
+  state->thread_count.Remove();
 }
 
-void ThreadPool::ThreadFunc() {
-  for (;;) {
-    // Wait until work is available or we are shutting down.
-    grpc_core::ReleasableMutexLock lock(&mu_);
-    if (!forking_ && !shutdown_ && callbacks_.empty()) {
-      // If there are too many threads waiting, then quit this thread
-      if (threads_waiting_ >= reserve_threads_) {
-        break;
-      }
-      threads_waiting_++;
-      cv_.Wait(&mu_);
-      threads_waiting_--;
-    }
-    // a fork could be initiated while the thread was waiting
-    if (forking_) return;
-    // Drain callbacks before considering shutdown to ensure all work
-    // gets completed.
-    if (!callbacks_.empty()) {
-      auto cb = std::move(callbacks_.front());
-      callbacks_.pop();
-      lock.Release();
-      cb();
-    } else if (shutdown_) {
+bool ThreadPool::Queue::Step() {
+  grpc_core::ReleasableMutexLock lock(&mu_);
+  // Wait until work is available or we are shutting down.
+  while (state_ == State::kRunning && callbacks_.empty()) {
+    // If there are too many threads waiting, then quit this thread.
+    // TODO(ctiller): wait some time in this case to be sure.
+    if (threads_waiting_ >= reserve_threads_) return false;
+    threads_waiting_++;
+    cv_.Wait(&mu_);
+    threads_waiting_--;
+  }
+  switch (state_) {
+    case State::kRunning:
       break;
-    }
+    case State::kShutdown:
+    case State::kForking:
+      if (!callbacks_.empty()) break;
```

Comment:
I think this change in behavior could cause a hang on fork or shutdown. If a callback queues more callbacks, `Queue::Add` will take those on even if state==kForking, and ThreadFunc will continue to execute until all the callbacks are exhausted. Instead, callback execution could be halted until the fork completes.

I believe it's straightforward to test. Create a callback that sleeps a bit and schedules its own next iteration, then exits when some post-fork condition is met. Then start a thread that calls PreFork (which I think will hang) and PostForkParent, then sets the post-fork condition. The whole thing should complete in < 5s conservatively (probably millis on an unloaded system).

Reply:
I'm having trouble convincing myself that such a scenario is legal.

Consider the following test:
```
TEST(ThreadPoolTest, CanScheduleAcrossFork) {
  ThreadPool p(1);
  absl::Notification callback_started;
  absl::Notification post_fork;
  absl::Notification callback_finished;
  p.Add([&callback_started, &post_fork, &callback_finished] {
    callback_started.Notify();
    post_fork.WaitForNotification();
    callback_finished.Notify();
  });
  callback_started.WaitForNotification();
  p.PrepareFork();
  p.PostforkParent();
  post_fork.Notify();
  callback_finished.WaitForNotification();
}
```

Such a test can never finish because it needs the fork to finish before it can end all pending callbacks.
We must end all pending callbacks prior to forking.

What you're proposing simply adds a sleep in the path, which would change this from deterministically failing to flakily failing (with a flake rate chosen by the sleep time).

Which is to say that I think the system must settle out prior to forking without recourse to knowing if the forking has finished or not - and if that's not the case we should fix the callback and not the thread pool.

Reply:
Fair point, that test is problematic, but the broader problem still exists. My suggestion is to either (A) queue but not run all new closures until the fork is complete, or (B) refuse accepting new closures until the fork completes. I think (B) is more problematic because closures will have to handle pool refusals when they should be able to ignore forks. Doing neither I believe guarantees the possibility of a never-ending fork.

```
void ScheduleChainedCallback(ThreadPool& pool, Promise& forked, Promise& done) {
  // do work, then ...
  if (forked.WaitWithTimeout(Duration::Zero())) {
    done.Set(true);
    return;
  }
  pool.Add([&] { ScheduleChainedCallback(pool, forked, done); });
}

TEST(ThreadPoolTest, ForkProceeds) {
  ThreadPool pool(1);
  grpc_event_engine::experimental::Promise<bool> forked(false);
  grpc_event_engine::experimental::Promise<bool> done(false);
  ScheduleChainedCallback(pool, forked, done);
  pool.PrepareFork();
  pool.PostforkParent();
  forked.Set(true);
  ASSERT_TRUE(done.Get());
}
```

You can replace the fork promise check with any real-life boundary condition appropriate for that work (i.e., keep doing this work until X, assuming a fork happens before X). I think that's a legitimate use case, one we've talked about with the poller interface in fact. Does that make sense?

Reply:
So my contention here is still: if that X wants to go forever, it needs to factor in wanting to fork as being the end of its forever.

The real life instance that I can think of for wanting to continually call a callback is for polling/timer checks, and that I think should be fork aware. I think other instances are the same, but would be happy to see a contra case.

Reply:
I can agree with making more things fork-aware as a solution. However that could add complexity to things that don't actually care if a fork happens (they do the same thing before and after regardless, but have to orchestrate an extra stop and start). That would be work done exclusively to prevent hangs due to a side effect of how the thread pool works, which seems like a leaky abstraction. Maybe it's a small set of edge cases, to your point. And we may find them if there are random hangs in pre-fork.

If the thread pool halts execution with callbacks still in queue, and some callback really wanted to do something pre-fork, its owner can opt in to being fork-aware. Otherwise the default expectation can be that callbacks are unbothered by forking.

Reply:
So I'm going to claim that we don't do forever chains of callbacks in gRPC (except in places that need to be fork aware anyway), and I think this is true for other things too - I can't think of a compelling case I've seen that would need forever rescheduled callbacks and forks and not be able to handle that.

That said, accidents happen and it'd be better to be able to diagnose them, so I've added some diagnostics to help find this when it does happen (and a test to ensure that it works).

Reply:
Ack. Let's go with your plan and see how it shakes out.